examples/t_factory.py

```diff
@@ -0,0 +1,94 @@
+import numpy as np
+
+from guppylang.decorator import guppy
+from guppylang.module import GuppyModule
+from guppylang.prelude.builtins import linst, py
+from guppylang.prelude.quantum import (
+    cz,
+    discard,
+    h,
+    measure,
+    quantum,
+    qubit,
+    rx,
+    rz,
+)
+
+module = GuppyModule("tfactory")
+module.load(quantum)
+
+phi = np.arccos(1 / 3)
+pi = np.pi
+
+
+@guppy(module)
+def ry(q: qubit, theta: float) -> qubit:
+    q = rx(q, py(pi / 2))
+    q = rz(q, theta + py(pi))
+    q = rx(q, py(pi / 2))
+    return rz(q, py(pi))
+
+
+# Preparation of approximate T state, from 2310.12106
+@guppy(module)
+def prepare_approx(q: qubit) -> qubit:
+    phi_ = py(phi)
+    pi_ = py(pi)
+
+    q = ry(q, phi_)
+    return rz(q, pi_ / 4.0)
+
+
+# The inverse of the [[5,3,1]] encoder in figure 3 of 2208.01863
+@guppy(module)
+def distill(
+    target: qubit, q0: qubit, q1: qubit, q2: qubit, q3: qubit
+) -> tuple[qubit, bool]:
+    """First argument is the target qubit which will be returned from the circuit.
+    Other arguments are ancillae, which should also be in an approximate T state.
+    Returns target qubit and a bool, which is true if the distillation succeeded.
+    """
+    q0, q1 = cz(q0, q1)
+    q2, q3 = cz(q2, q3)
+    target, q0 = cz(target, q0)
+    q1, q2 = cz(q1, q2)
+    target, q3 = cz(target, q3)
+    # Measuring gives false for success, true for failure.
+    # We check for all falses to say whether distillation succeeded.
+    bits = [not (measure(h(q))) for q in [q0, q1, q2, q3]]
+    success = True
```

Comment:
Can this be just `success = all(bits)`? Hmmm, maybe that's not in guppy yet - if so, suggest commenting that, I bet it'll go in there at some point :-)

Reply:
That's what I did on my first attempt, but it's not supported by guppy yet!